src/ray/gcs/store_client/redis_store_client.cc

```diff
@@ -244,16 +218,115 @@ Status RedisStoreClient::AsyncMultiGet(
+    if (added) {
+      queue_added++;
+    }
+    if (added) {
+      // There is no in-flight requests for this key which means
+      // in the callback the queue won't be poped. So just push
+      // an entry as placeholder to indicate there is pending
+      // requests.
+      op_iter->second.push(nullptr);
+    } else {
+      op_iter->second.push(send_request);
+    }
+  }
+  return queue_added;
+}
+
+std::vector<std::function<void()>> RedisStoreClient::PopFromSendingQueue(
+    const std::vector<std::string> &keys) {
+  std::vector<std::function<void()>> send_requests;
+  for (const auto &key : keys) {
+    auto [op_iter, added] =
+        pending_redis_request_by_key_.emplace(key, std::queue<std::function<void()>>());
+    RAY_CHECK(added == false) << "Pop from a queue doesn't exist: " << key;
+    op_iter->second.pop();
+    if (op_iter->second.empty()) {
+      pending_redis_request_by_key_.erase(op_iter);
+    } else {
+      send_requests.emplace_back(std::move(op_iter->second.front()));
+    }
+  }
+  return send_requests;
+}
+
+void RedisStoreClient::SendRedisCmd(std::vector<std::string> keys,
+                                    std::vector<std::string> args,
+                                    RedisCallback redis_callback) {
+  RAY_CHECK(!keys.empty());
+  // The number of keys that's ready for this request.
+  // For a query reading or writing multiple keys, we need a counter
+  // to check whether all existing requests for this keys have been
+  // processed.
+  auto num_ready_keys = std::make_shared<std::atomic<size_t>>(0);
+  std::function<void()> send_redis = [this,
+                                      num_ready_keys = num_ready_keys,
+                                      keys,
+                                      args = std::move(args),
+                                      redis_callback =
+                                          std::move(redis_callback)]() mutable {
+    {
+      absl::MutexLock lock(&mu_);
+      *num_ready_keys += 1;
+      RAY_CHECK(*num_ready_keys <= keys.size());
+      // There are still pending requets for these keys.
+      if (*num_ready_keys != keys.size()) {
+        return;
+      }
+    }
+    // Send the actual request
+    auto cxt = redis_client_->GetShardContext("");
+    RAY_CHECK_OK(cxt->RunArgvAsync(
+        std::move(args),
+        [this, keys = std::move(keys), redis_callback = std::move(redis_callback)](
+            auto reply) {
+          std::vector<std::function<void()>> requests;
+          {
+            absl::MutexLock lock(&mu_);
+            requests = PopFromSendingQueue(keys);
+          }
+          for (auto &request : requests) {
+            request();
+          }
+          if (redis_callback) {
+            redis_callback(reply);
+          }
+        }));
+  };
+
+  {
+    absl::MutexLock lock(&mu_);
+    auto keys_ready = PushToSendingQueue(keys, send_redis);
+    *num_ready_keys += keys_ready;
+    // We should be able to file it directly.
```

Comment:
nit: file -> fire

worth add a bit more comments:

"as part of the contract, if the every queue is empty for each key this request depends on, we should fire the request immediately".